homeassistant/components/mqtt/discovery.py

```diff
@@ -51,47 +51,64 @@
-            payload[CONF_STATE_TOPIC] = '{}/{}/{}{}/state'.format(
-                discovery_topic, component, '%s/' % node_id if node_id else '',
-                object_id)
-
-        if ALREADY_DISCOVERED not in hass.data:
-            hass.data[ALREADY_DISCOVERED] = set()
-
         # If present, the node_id will be included in the discovered object id
         discovery_id = '_'.join((node_id, object_id)) if node_id else object_id
 
-        discovery_hash = (component, discovery_id)
-        if discovery_hash in hass.data[ALREADY_DISCOVERED]:
-            _LOGGER.info("Component has already been discovered: %s %s",
-                         component, discovery_id)
-            return
-
-        hass.data[ALREADY_DISCOVERED].add(discovery_hash)
+        if ALREADY_DISCOVERED not in hass.data:
+            hass.data[ALREADY_DISCOVERED] = {}
 
-        _LOGGER.info("Found new component: %s %s", component, discovery_id)
+        discovery_hash = (component, discovery_id)
 
-        await async_load_platform(
-            hass, component, platform, payload, hass_config)
+        if payload:
+            # Add component
+            try:
+                payload = json.loads(payload)
+            except ValueError:
+                _LOGGER.warning("Unable to parse JSON %s: %s",
+                                object_id, payload)
+                return
+
+            payload = dict(payload)
+            platform = payload.get(CONF_PLATFORM, 'mqtt')
+            if platform not in ALLOWED_PLATFORMS.get(component, []):
+                _LOGGER.warning("Platform %s (component %s) is not allowed",
+                                platform, component)
+                return
+
+            payload[CONF_PLATFORM] = platform
+            if CONF_STATE_TOPIC not in payload:
+                payload[CONF_STATE_TOPIC] = '{}/{}/{}{}/state'.format(
+                    discovery_topic, component,
+                    '%s/' % node_id if node_id else '', object_id)
+
+            if discovery_hash in hass.data[ALREADY_DISCOVERED]:
+                _LOGGER.info("Component has already been discovered: %s %s",
+                             component, discovery_id)
+                return
+
+            hass.data[ALREADY_DISCOVERED][discovery_hash] = None
+            payload[ATTR_DISCOVERY_HASH] = discovery_hash
+
+            _LOGGER.info("Found new component: %s %s", component, discovery_id)
+
+            await async_load_platform(
+                hass, component, platform, payload, hass_config)
+
+        else:
+            # Remove component
+            if discovery_hash in hass.data[ALREADY_DISCOVERED]:
+                _LOGGER.info("Removing component: %s %s",
+                             component, discovery_id)
+
+                if hass.data[ALREADY_DISCOVERED][discovery_hash] is not None:
+                    entity = hass.data[ALREADY_DISCOVERED][discovery_hash]
+                    hass.states.async_remove(entity.entity_id)
```

Comment:
We shouldn't let the discovery code have access to entity objects.

MQTT entities created from a discovery hash should subscribe to `mqtt_discovery_updated` signal with [the dispatcher](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/dispatcher.py) inside [the `async_added_to_hass` hook](https://developers.home-assistant.io/docs/en/entity_index.html#async-added-to-hass). 

That way we can also push name updates or even subscribe to other channels if we would want.

The code would turn into something like:

```python
if discovery_hash in hass.data[ALREADY_DISCOVERED]:
    hass.helpers.dispatcher. async_dispatcher_send('mqtt_discovery_updated', payload)
else:
    await async_load_platform(
        hass, component, platform, payload, hass_config)
```

Reply:
OK, sounds reasonable. Do you suggest to add subscription to `mqtt_discovery_updated` signal in each object type: light, switch, fan etc or add new class similar to MqttAvailability?